llvm/lib/IR/DebugInfo.cpp

```diff
@@ -375,6 +375,32 @@ bool DebugInfoFinder::addScope(DIScope *Scope) {
   return true;
 }
 
+// Recursively handle DILocations in followup metadata etc.
+static Metadata *updateLoopMetadataDebugLocationsRecursive(
+    Metadata *MetadataIn, function_ref<Metadata *(Metadata *)> Updater) {
+  const MDNode *M = dyn_cast_or_null<MDNode>(MetadataIn);
+  // The loop metadata options should start with a MDString.
+  if (!M || M->getNumOperands() < 1 || !isa<MDString>(M->getOperand(0)))
+    return nullptr;
+
+  bool Updated = false;
+  SmallVector<Metadata *, 4> MDs{M->getOperand(0)};
+  for (Metadata *MD : llvm::drop_begin(M->operands())) {
+    if (!MD) {
+      MDs.push_back(nullptr);
+    } else if (Metadata *NewMD =
+                   updateLoopMetadataDebugLocationsRecursive(MD, Updater)) {
+      MDs.push_back(NewMD);
+      Updated = true;
+    } else if (Metadata *NewMD = Updater(MD)) {
+      MDs.push_back(NewMD);
+      Updated |= NewMD != MD;
+    }
+  }
+
+  return Updated ? MDNode::get(M->getContext(), MDs) : nullptr;
```

Comment:
```suggestion
  return Updated ? MDNode::get(M->getContext(), MDs) : MetadataIn;
```
[suggestion] I would return the old MD if there is no change. The caller can still check whether it is the same, but is not required to.

Reply:
Thanks! I've changed it to return the input node if not making changes, and then I could collapse the if-else logic a bit.